### PR TITLE
CIVIPLMMSR-418: Add Upgrader To Fix Faulty Contributions

### PR DESCRIPTION
**Pull request by shahrukh-compuco**
## Overview
This pr adds an upgrader to fix the faulty contribution records that does not have any line item attached to it.

## Technical Details
Such contributions gets created when a membership record is first created without a recurring contribution and then renewed with a contribution in CiviCRM. Systems who are migrated from another CRM and don’t have recurring contributions associated with memberships will likely experience this issue.

The root cause of this issue is renewing memberships without prior contribution (or if CiviCRM cannot find the contribution), the contribution created for the membership renewal will not have the appropriate line items.

This pr fixes these records by creating appropriate financial transactions, line items and financial items as well. Although it is possible that in past there were multiple transactions were there for a faulty contribution so we will just create the financial transaction records for the last transaction.

### CODE REVIEW

File: Civi/Financeextras/Service/IncompleteContributionFixService.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
+<?php
+
+namespace Civi\Financeextras\Service;
```
Comment (erawat) [inline Civi/Financeextras/Service/IncompleteContributionFixService.php:3]: @shahrukh-compuco  I think the name space should be Civi\Financeextras\Upgrade;  We only want to run this one time.

File: Civi/Financeextras/Service/IncompleteContributionFixService.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    if ($createFinancialTrxn) {
+      civicrm_api3('FinancialTrxn', 'create', [
+        'to_financial_account_id' => 'Payment Processor Account',
+        'trxn_date' => $contribution['receive_date'],
+        'total_amount' => $contribution['total_amount'],
+        'fee_amount' => $contribution['fee_amount'],
+        'net_amount' => $contribution['net_amount'],
+        'currency' => $contribution['currency'],
+        'is_payment' => 1,
+        'trxn_id' => $contribution['trxn_id'],
```
Comment (erawat) [inline Civi/Financeextras/Service/IncompleteContributionFixService.php:85]: @shahrukh-compuco  Where does $contribution get the trxn_id from?  

The requirement was that we need to get the real Stripe transaction ID. 

https://compucorp.atlassian.net/wiki/spaces/CN/pages/5012881470/CIVIPLMMSR-314+-+Successful+Stripe+transactions+are+marked+as+pending+in+CiviCRM#:~:text=Record%20the%20Stripe%20transaction%20ID%20get%20from%20Stripe%20for%20the%20contribution%20ID%20in%20the%20financial%20transaction%20and%20on%20the%20contribution%20in%20the%20relevant%20places.

**Review by erawat — CHANGES_REQUESTED** [review]
@shahrukh-compuco  @olayiwola-compucorp   The upgrader should be part of Stripe extension, as we need to record Stripe transaction ID to the payment.

**Comment by shahrukh-compuco** [comment]
> @shahrukh-compuco @olayiwola-compucorp The upgrader should be part of Stripe extension, as we need to record Stripe transaction ID to the payment.

yes I and @olayiwola-compucorp discussed this but since this is not just related to stripe so decided to keep it in financeextras instead

File: Civi/Financeextras/Service/IncompleteContributionFixService.php (hunk 1 above)
Comment (shahrukh-compuco) [reply Civi/Financeextras/Service/IncompleteContributionFixService.php:3]: not sure on this as the namespace imo needs to reflect the directory structure for autoloading purposes and this class resides in services folder

File: Civi/Financeextras/Service/IncompleteContributionFixService.php (hunk 2 above)
Comment (shahrukh-compuco) [reply Civi/Financeextras/Service/IncompleteContributionFixService.php:85]: yes if it was a stripe payment then the transaction id is from stripe in contribution table

Comment (erawat) [reply Civi/Financeextras/Service/IncompleteContributionFixService.php:85]: @shahrukh-compuco  I am confused, how did you get Stripe transaction ID it was never been recorded? Don't you have to fetch from Stripe?

File: Civi/Financeextras/Service/IncompleteContributionFixService.php (hunk 1 above)
Comment (erawat) [reply Civi/Financeextras/Service/IncompleteContributionFixService.php:3]: @shahrukh-compuco  Shouldn't you create a new directory?

**Review by erawat — CHANGES_REQUESTED** [review]

File: Civi/Financeextras/Service/IncompleteContributionFixService.php (hunk 2 above)
Comment (shahrukh-compuco) [reply Civi/Financeextras/Service/IncompleteContributionFixService.php:85]: No in the scenario that we are trying to fix contribution table already has the transaction id correct me @olayiwola-compucorp if I am wrong but the spec also says to use the transaction id from contribution https://compucorp.atlassian.net/wiki/spaces/CN/pages/5012881470/CIVIPLMMSR-314+-+Successful+Stripe+transactions+are+marked+as+pending+in+CiviCRM#3.-Create-Missing-Financial-Transactions

**Comment by shahrukh-compuco** [comment]
@erawat the spec said to create the class as a service so I am not sure that it will be a good idea to have two directories for services

File: Civi/Financeextras/Service/IncompleteContributionFixService.php (hunk 2 above)
Comment (olayiwola-compucorp) [reply Civi/Financeextras/Service/IncompleteContributionFixService.php:96]: @shahrukh-compuco @erawat For most of the cases that were reported, the contribution had the correct Stripe transaction ID. but if we want to cover edge cases where the contribution transaction ID is empty, we may have to move to stripe.

File: tests/phpunit/Civi/Financeextras/Service/IncompleteContributionFixServiceTest.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      'membership_type_id' => $membershiptType['id'],
+      'join_date' => $today,
+      'start_date' => $today,
+      'financial_type_id' => 'Member Dues',
+      'skipLineItem' => 1,
+    ]);
+
+    $createContributionQuery = "
+    INSERT INTO civicrm_contribution
+    (contact_id, financial_type_id, receive_date, total_amount, payment_instrument_id, trxn_id, currency, contribution_status_id)
```
Comment (olayiwola-compucorp) [inline tests/phpunit/Civi/Financeextras/Service/IncompleteContributionFixServiceTest.php:36]: Set is_pay_later to  0 @shahrukh-compuco

Comment (shahrukh-compuco) [reply tests/phpunit/Civi/Financeextras/Service/IncompleteContributionFixServiceTest.php:36]: @olayiwola-compucorp by default it is 0 in db table

File: tests/phpunit/Civi/Financeextras/Service/IncompleteContributionFixServiceTest.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    $contributionId = CRM_Core_DAO::singleValueQuery('SELECT LAST_INSERT_ID()');
+
+    $createMembershipPaymentQuery = "
+    INSERT INTO civicrm_membership_payment
+    (contribution_id, membership_id)
+    VALUES
+    ({$contributionId}, {$membership['id']});";
+    \CRM_Core_DAO::executeQuery($createMembershipPaymentQuery);
+
+    $service = new IncompleteContributionFixService();
```
Comment (olayiwola-compucorp) [inline tests/phpunit/Civi/Financeextras/Service/IncompleteContributionFixServiceTest.php:49]: A unit test should follow the following pattern:
Arrange -> Act -> Assert

In the Arrange step, which is where you're creating the contributions, you also need to confirm the invalid contribution counts, and that the necessary financial records are missing.

Then act by executing the `IncompleteContributionFixService` class as you're doing

Then you assert the count of the invalid contributions and necessary records that needs to be created have been created with the correct values.

Also, is there any reason we chose not to use API4 or fabricator class

Comment (shahrukh-compuco) [reply tests/phpunit/Civi/Financeextras/Service/IncompleteContributionFixServiceTest.php]: changed

**Review by olayiwola-compucorp — APPROVED** [review]

File: Civi/Financeextras/Service/IncompleteContributionFixService.php [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+      'entity_table' => 'civicrm_membership',
+      'entity_id' => $membership['membership_id'],
+      'contribution_id' => $contribution['id'],
+      'price_field_id' => $priceFieldValue['price_field_id'],
+      'price_field_value_id' => $priceFieldValue['id'],
+      'label' => $priceFieldValue['name'],
+      'qty' => 1,
+      'unit_price' => $contribution['total_amount'],
+      'line_total' => $contribution['total_amount'],
+      'financial_type_id' => 'Member Dues',
```
Comment (erawat) [inline Civi/Financeextras/Service/IncompleteContributionFixService.php:126]: @shahrukh-compuco  can we get financial type from the contribution?

**Review by erawat — CHANGES_REQUESTED** [review]

File: Civi/Financeextras/Service/IncompleteContributionFixService.php (hunk 5 above)
Comment (shahrukh-compuco) [reply Civi/Financeextras/Service/IncompleteContributionFixService.php]: done

**Review by erawat — APPROVED** [review]
